src/CMakeLists.txt

```diff
@@ -135,136 +135,131 @@ if(DLAF_WITH_PRECOMPILED_HEADERS)
 endif()
 
-# Define DLAF's CORE library
-add_library(dlaf.core
-  OBJECT
-    communication/communicator_impl.cpp
-    communication/communicator.cpp
-    communication/communicator_grid.cpp
-    communication/datatypes.cpp
-    init.cpp
-    matrix/distribution.cpp
-    matrix/layout_info.cpp
-    matrix/tile.cpp
-    matrix.cpp
-    matrix_mirror.cpp
-    memory/memory_view.cpp
-    memory/memory_chunk.cpp
-)
+# Helper function for creating sub-libraries that libDLAF consists of. It
+# creates two variants of the library: one object library with the name
+# dlaf.${name}_object and one static/shared library with the name dlaf.${name}.
+# The object libraries are only meant for creating libDLAF. The static/shared
+# libraries are meant for internal use in tests and miniapps.
+function(DLAF_addSublibrary name)
+  set(options "")
+  set(oneValueArgs "")
+  set(multiValueArgs SOURCES GPU_SOURCES COMPILE_OPTIONS LIBRARIES)
+  cmake_parse_arguments(DLAF_ASL "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
 
-if(DLAF_WITH_GPU)
-  set(dlaf_core_gpu_source_files
-    cusolver/assert_info.cu
-    cusolver/stedc.cu
-    lapack/gpu/lacpy.cu
-    lapack/gpu/laset.cu
-  )
-  target_sources(dlaf.core PRIVATE ${dlaf_core_gpu_source_files})
-  if(DLAF_WITH_HIP)
-    set_source_files_properties(${dlaf_core_gpu_source_files} PROPERTIES LANGUAGE HIP)
+  set(object_lib_name dlaf.${name}_object)
+  set(lib_name dlaf.${name})
+
+  add_library(${object_lib_name} OBJECT ${DLAF_ASL_SOURCES})
+  if(DLAF_WITH_GPU)
+    target_sources(${object_lib_name} PRIVATE ${DLAF_ASL_GPU_SOURCES})
+    if(DLAF_WITH_HIP)
+      set_source_files_properties(${DLAF_ASL_GPU_SOURCES} PROPERTIES LANGUAGE HIP)
+    endif()
   endif()
-endif()
 
-target_compile_options(dlaf.core PRIVATE $<$<COMPILE_LANG_AND_ID:CUDA,NVIDIA>:--extended-lambda>)
-target_link_libraries(dlaf.core PUBLIC dlaf.prop)
-target_add_warnings(dlaf.core)
-DLAF_addPrecompiledHeaders(dlaf.core)
+  target_compile_options(${object_lib_name} PRIVATE ${DLAF_ASL_COMPILE_OPTIONS})
+  target_link_libraries(${object_lib_name} PRIVATE dlaf.prop)
+  target_add_warnings(${object_lib_name})
+  DLAF_addPrecompiledHeaders(${object_lib_name})
+  add_library(${lib_name} $<TARGET_OBJECTS:${object_lib_name}>)
+  target_link_libraries(${lib_name} PUBLIC dlaf.prop)
+  target_link_libraries(${lib_name} PUBLIC ${DLAF_ASL_LIBRARIES})
+endfunction()
+
+# Define DLAF's CORE library
+DLAF_addSublibrary(
+  core
+  SOURCES communication/communicator_impl.cpp
+          communication/communicator.cpp
+          communication/communicator_grid.cpp
+          communication/datatypes.cpp
+          init.cpp
+          matrix/distribution.cpp
+          matrix/layout_info.cpp
+          matrix/tile.cpp
+          matrix.cpp
+          matrix_mirror.cpp
+          memory/memory_view.cpp
+          memory/memory_chunk.cpp
+  GPU_SOURCES cusolver/assert_info.cu cusolver/stedc.cu lapack/gpu/lacpy.cu lapack/gpu/laset.cu
```

Comment:
Additional comment about `GPU_SOURCES` to hopefully clarify its purpose (thanks @albestro for bringing it up!): `GPU_SOURCES` contains `.cu` files (not `.cpp` files that contain GPU functionality). The main purpose for its existence is so that we can call `set_source_files_properties(${DLAF_ASL_GPU_SOURCES} PROPERTIES LANGUAGE HIP)` on them. If this weren't required all files could be in `SOURCES`. An alternative solution would be to look at the file extensions and automatically detect the `.cu` files.

I'm also commenting here in case someone things a different solution or name would be more appropriate here. After my explanations @albestro claims that it's clear, but I'm nonetheless open to changing it if you think it can be made clearer.